babel.config.js

```diff
@@ -0,0 +1,41 @@
+const config = {
+	presets: [
+		[
+			'@babel/env',
+			{
+				modules: 'commonjs',
+				useBuiltIns: 'entry',
+				shippedProposals: true, // allows es7 features like Promise.prototype.finally
+			},
+		],
+		'@babel/preset-react'
+	],
+	plugins: [
+		'@babel/plugin-transform-member-expression-literals',
+		'@babel/plugin-transform-react-jsx',
+		'@babel/plugin-transform-runtime',
+
+		// Stage 1
+		'@babel/plugin-proposal-export-default-from',
+		'@babel/plugin-proposal-logical-assignment-operators',
```

Comment:
These plugins for unstable proposals were removed in my version of the branch from December 2018. Is the re-adding intentional, or just an unsuccessful rebase?

Reply:
Unsuccessful rebase. 👍  So I'll remove all `plugin-proposal-*` ones?

Reply:
And suppose also `shippedProposal` from `@babel/preset-env` config?

Reply:
This is the config that's in my working copy, slightly amended:
```js
const config = {
  presets: [
    [
      '@babel/env',
      {
        modules: 'commonjs',
        useBuiltIns: 'entry',
        corejs: 2,
      },
    ],
    '@babel/react'
  ],
  plugins: [
    '@babel/plugin-proposal-class-properties',
    '@babel/transform-runtime',
    'transform-member-expression-literals',
    'transform-property-literals',
    'transform-reserved-words',
  ]
};
```

The `shippedProposals: true` option is not needed today. It means "support features that shipped in browsers (i.e., are supported natively), but were not yet approved as part of ES2018". Once the ES2018 standard was approved and Babel updated to reflect the standard, the set became empty.

It will get populated again as the ES2019 standard is finalized, I suppose. Then we can choose to include `shippedProposals` again if the new features are interesting.

The `transform-member-expression-literals` plugins are needed for ES3 compatibility with old IEs that Jetpack still keeps (?). `test.catch` is syntax error in ES3, because `catch` is reserved word. It's transformed to `test['catch']`.

Reply:
Thanks or elaborating @jsnajdr !

Reply:
I started setting those up but I'm not sure what's the difference between `@babel/react` vs `@babel/react-preset` and which we need.

Also couldn't find plugin for `transform-reserved-words` but `@babel/plugin-transform-reserved-words` exists.

Additionally, I started pinning versions and removing unneeded stuff from `package.json`:
```
    "@babel/core": "7.4.0",
    "@babel/plugin-proposal-class-properties": "7.4.0",
    "@babel/plugin-syntax-dynamic-import": "7.2.0",
    "@babel/plugin-syntax-import-meta": "7.2.0",
    "@babel/plugin-transform-member-expression-literals": "7.2.0",
    "@babel/plugin-transform-react-jsx": "7.3.0",
    "@babel/plugin-transform-reserved-words": "7.2.0",
    "@babel/plugin-transform-runtime": "7.4.0",
    "@babel/preset-env": "7.4.2",
    "@babel/preset-react": "7.0.0",
    "@babel/register": "7.4.0",
    "babel-eslint": "10.0.1",
    "babel-loader": "8.0.5",
    "babel-plugin-transform-member-expression-literals": "6.9.4",
    "babel-plugin-transform-property-literals": "6.9.4",
```

As well removed these Babel options overrides from Webpack config entirely:

https://github.com/Automattic/jetpack/blob/ff51438bc33c67be344edad38cd9b48e225294cb/webpack.config.js#L28-L55


...but since the build is broken with these changes and I'm getting a bit too deep into something I shouldn't really be spending time on too much and I'm not too familiar with Babel anyway, I'll give up and leave this update for someone else. :-)

Reply:
@jsnajdr if you suspect I made some broken chagnes with rebase and you have locally your changes still, I'd suggest you push that to either here by `--force` (in which case we just need to rebase again) or to a different branch name.

Reply:
> I started setting those up but I'm not sure what's the difference between @babel/react vs @babel/react-preset and which we need.

Babel allows you to specify various smart shorthands in the config file and expands them when looking up the corresponding NPM package:
- `plugins: [ 'foo' ]` -> `babel-plugin-foo`
- `presets: [ 'bar' ]` -> `babel-preset-bar`
- `presets: [ '@babel/react' ]` -> `@babel/preset-react`
- `plugins: [ '@wordpress' ]` -> `@wordpress/babel-plugin`
- `presets: [ '@wordpress/default' ]` -> `@wordpress/babel-preset-default`

I think that suffixes (`foo` -> `foo-babel-plugin`) work, too.